### ISSUE TITLE
After running for days FastAPI + ultalytics.YOLO goes blind

### ISSUE THREAD

**Issue by amarder** [ultralytics/ultralytics#23285]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi All,

I'm running into a bug that is hard for me to reproduce. I'm using FastAPI to deploy YOLOv11 on a server with a GPU. Everything runs smoothly for a while, but something happens and the model goes blind, `yolo.predict(image)` starts returning empty lists on images that had predictions previously. Has anyone seen behavior like this before? When I restart the server everything works again.

Thanks, Andrew

**Comment by UltralyticsAssistant** [comment]
👋 Hello @amarder, thank you for your interest in Ultralytics 🚀! This is an automated response to help us triage your issue quickly—an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) to help us debug it. Since this appears intermittent (after days of FastAPI serving the model), an MRE is especially important 🙏.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

To help us investigate the “goes blind after days” behavior 🕵️, please include in your MRE:
-Exact `ultralytics` version, Python version, PyTorch version, and GPU/CUDA info
-A minimal FastAPI snippet showing how the `YOLO` model is instantiated (global vs per-request) and how `predict()` is called
-A few sample inputs (or a description of their shape/type/path) and the exact returned output structure when it “fails”
-Any logs leading up to the failure and whether memory usage changes over time

**Comment by glenn-jocher** [comment]
This usually ends up being a concurrency/thread-safety issue when a single shared `YOLO` instance is hit by multiple FastAPI requests over time, so please try either one model per worker process (recommended) or wrap inference with a lock as shown in the [thread-safe inference guide](https://docs.ultralytics.com/guides/yolo-thread-safe-inference/); if you can also paste `yolo checks` output plus a truly minimal repro (ideally using `yolo26n.pt` + a public image) following our [Minimum Reproducible Example guide](https://docs.ultralytics.com/help/minimum-reproducible-example/), we can help narrow down whether it’s an environment/version conflict or a state/race condition.

```python
from ultralytics import YOLO
from ultralytics.utils import ThreadingLocked

model = YOLO("yolo26n.pt")

@ThreadingLocked()
def locked_predict(img):
    return model.predict(source=img, verbose=False)
```

**Comment by amarder** [comment]
Thanks for the help Glenn! I'll work on a minimum reproducible example (I am using `ultralytics.utils.ThreadingLocked`, initializing a new model for every prediction would be too slow). The code for [ThreadingLocked](https://github.com/ultralytics/ultralytics/blob/3f66a283e35247d3a4a2a74a34d2c0b1ab688c51/ultralytics/utils/__init__.py#L473-L503) is nice and clear. Looks like it's time for me to learn more about FastAPI and Python threads.

**Comment by glenn-jocher** [comment]
Good call using `ThreadingLocked`, but in FastAPI the other common gotcha is your server process model: if you’re running multiple `uvicorn`/Gunicorn workers, each worker should load its own `YOLO` instance (and if you’re using Gunicorn with CUDA, avoid `--preload` since forking after CUDA init can cause weird long-run behavior); can you paste the exact command you use to start the server (incl. `--workers`/Gunicorn flags) and any logs from the time it starts returning empty results? The relevant concurrency note is also called out in the [predict mode docs (thread-safe inference section)](https://docs.ultralytics.com/modes/predict/).

```bash
uvicorn your_module:app --host 0.0.0.0 --port 8000 --workers 1
```

**Comment by glenn-jocher** [comment]
We've got FastAPI powering our demo here also running for a few days, so far so good :)
https://platform.ultralytics.com/glenn-jocher/yolo26/yolo26n?tab=predict

<img width="1557" height="1337" alt="Image" src="https://github.com/user-attachments/assets/75337cb2-3db2-4a53-8acb-c14b39a6ef15" />

**Comment by amarder** [comment]
Thanks Glenn!

Here's a minimal example of what my current Python code looks like:

```python
from fastapi import FastAPI, File, UploadFile
from ultralytics import YOLO
from ultralytics.utils import ThreadingLocked
from PIL import Image
import io
import torch

app = FastAPI(title="YOLO Prediction API")
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
model = YOLO("yolo26n.pt").to(device)

@ThreadingLocked()
def _model_predict(image: Image.Image) -> list[dict]:
    results = model.predict(image)    
    return [result.summary() for result in results]

@app.post("/predict")
async def predict(file: UploadFile = File(...)):
    """Upload an image and get YOLO predictions."""
    contents = await file.read()
    image = Image.open(io.BytesIO(contents))
    return _model_predict(image)
```

Ideally, I'd like to make predictions leveraging the GPU for speed (hence the `.to(device)` call). I've been running the app with multiple workers like this:

```
uvicorn your_module:app --host 0.0.0.0 --port 8000 --workers 4
```

Hoping that uvicorn would keep the workers separate, but maybe that's wrong.

**Comment by glenn-jocher** [comment]
`uvicorn --workers 4` gives you 4 separate *processes* (not threads), so each worker has its own `YOLO` instance and its own `ThreadingLocked` lock; on a single GPU this usually just means 4 independent CUDA contexts fighting for the same device over days, so the most stable setup is typically 1 worker per GPU and then serialize/queue requests inside that worker (your `ThreadingLocked` approach is fine; see the [ThreadingLocked API reference](https://docs.ultralytics.com/reference/utils/__init__/)).

One practical tweak I’d recommend is to drop `model = YOLO(...).to(device)` and instead pin the device at call-time (and force RGB), e.g.:

```python
from fastapi import FastAPI, File, UploadFile, Request
from ultralytics import YOLO
from ultralytics.utils import ThreadingLocked
from PIL import Image
import io

app = FastAPI()
lock = ThreadingLocked()

@app.on_event("startup")
def _startup():
    app.state.model = YOLO("yolo26n.pt")

@lock
def _model_predict(model, image):
    return model.predict(source=image, device=0, verbose=False)

@app.post("/predict")
async def predict(request: Request, file: UploadFile = File(...)):
    image = Image.open(io.BytesIO(await file.read())).convert("RGB")
    results = _model_predict(request.app.state.model, image)
    return [r.summary() for r in results]
```

If it still “goes blind”, please paste `yolo checks` output from the same runtime plus any warnings/errors logged around the time detections stop (text, not screenshots), since that’s often enough to spot a CUDA/env conflict or an upstream decode/input issue.

**Comment by amarder** [comment]
Thanks Glenn, this is super helpful! I'll report back on how things go...

**Comment by glenn-jocher** [comment]
Sounds good—when you test, try running a single GPU worker *and* cap in-flight requests so they queue cleanly (this avoids slow buildup/edge cases from many concurrent connections even if `ThreadingLocked` serializes inference):

```bash
uvicorn your_module:app --host 0.0.0.0 --port 8000 --workers 1 --limit-concurrency 1
```

If it still “goes blind,” please paste the full text logs from the moment it happens (including image mode/size and `len(results[0].boxes)`), plus `!yolo checks` from that same running process; that usually makes these long-run issues diagnosable, and the background rationale is in the [thread-safe inference guide](https://docs.ultralytics.com/guides/yolo-thread-safe-inference/).

**Comment by amarder** [comment]
Hi Glenn,

I'm looking through the source code for `YOLO.predict` and I'm having a hard time seeing how `model.predict(source=image, device=0)` makes sure the computation happens on the GPU. The source code is here: https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/model.py#L515-L536

Any tips to understand how it works? Claude Opus 4.5 thinks both the model and the input tensor need to be moved to the GPU like this:

```python
import torch


class Wrapper:

    def __init__(self, model: torch.nn.Module):
        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        # 1. Move the model to GPU
        self.model = model.to(self.device)

    def predict(self, input_tensor: torch.Tensor) -> torch.Tensor:
        # 2. Move input tensors to GPU
        input_tensor = input_tensor.to(self.device)
        with torch.no_grad():
            # Return the prediction of the model
            return self.model(input_tensor)


if __name__ == "__main__":
    # Create a simple logistic regression model (2 inputs -> 1 output with sigmoid)
    model = torch.nn.Sequential(
        torch.nn.Linear(2, 1),
        torch.nn.Sigmoid(),
    )

    # Create a sample input tensor (batch of 5 samples, each with 2 features)
    input_tensor = torch.randn(5, 2)

    # Wrap the model and make a prediction
    wrapper = Wrapper(model)
    output = wrapper.predict(input_tensor)

    print(f"Device used: {wrapper.device}")
    print(f"Input shape: {input_tensor.shape}")
    print(f"Output shape: {output.shape}")
    print(f"Output (probabilities):\n{output}")
```

**Comment by glenn-jocher** [comment]
In raw PyTorch you do need to move both model + tensors, but Ultralytics handles that inside the predictor: passing `device=0` into `model.predict()` sets the predictor’s device, `BasePredictor.setup_model()` selects/builds the backend on that device, and `BasePredictor.preprocess()` moves the input tensor via `im = im.to(self.device)` before inference (see the [BasePredictor reference](https://docs.ultralytics.com/reference/engine/predictor/) and the [select_device reference](https://docs.ultralytics.com/reference/utils/torch_utils/)); you can sanity-check at runtime with:

```python
results = model.predict(source=image, device=0, verbose=False)
print(model.predictor.device)
```
